libmobilecoin/src/common/boundary.rs

```diff
@@ -52,11 +52,25 @@ where
     })
 }
 
-fn ffi_boundary_impl<R>(
-    f: impl (FnOnce() -> Result<R, LibMcError>) + UnwindSafe,
-) -> Result<R, LibMcError> {
+fn ffi_boundary_impl<R>(f: impl (FnOnce() -> Result<R, LibMcError>)) -> Result<R, LibMcError> {
     // Run f within catch_unwind
-    catch_unwind(f)
+    //
+    // Note: this is using AssertUnwindSafe because some of the TransactionBuilder
+    // types are abstracted behind Box<dyn ... + Send + Sync>, but rust does not
+    // allow such types to be UnwindSafe because they may exhibit interior
+    // mutability. OTOH if we do not put + Send + Sync, then it is illegal to
+    // put TransactionBuilder behind a Mutex, which prevents the android
+    // bindings from building.
+    //
+    // The reason we use Box<dyn + ...> at all is to avoid making everything
+    // a generic parameter of transaction builder, which multiplies the number of
+    // types that might have to have cross-language bindings.
+    //
+    // UnwindSafe is too restrictive -- the goal of UnwindSafe is that if a panic is
+    // caught, we cannot "easily" observe a broken invariant. However, the only
+    // thing we actually need at an ffi boundary is to prevent unwinding across
+    // stackframes into swift etc.
+    catch_unwind(AssertUnwindSafe(f))
```

Comment:
@voloshyn @mobilecoinfoundation/sdk Adam Mork, 

let me know if you guys agree with the changes in this file, I think this is the right thing to do (remove the `UnwindSafe` bound that was on all of these functions), I have tried to explain why I did this.

If I don't do this, I get build errors like this:

```
error[E0277]: the type `(dyn MemoBuilder + Send + Sync + 'static)` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   --> libmobilecoin/src/transaction.rs:343:5
    |
343 |     ffi_boundary_with_error(out_error, || {
    |     ^^^^^^^^^^^^^^^^^^^^^^^ `(dyn MemoBuilder + Send + Sync + 'static)` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
    | 
   ::: libmobilecoin/src/common/boundary.rs:38:15
    |
38  | pub(crate) fn ffi_boundary_with_error<R, I>(
    |               ----------------------- required by a bound in this
39  |     out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
40  |     f: impl (FnOnce() -> Result<R, LibMcError>) + UnwindSafe,
    |                                                   ---------- required by this bound in `boundary::ffi_boundary_with_error`
    |
    = help: within `std::option::Option<TransactionBuilder<FogResolver>>`, the trait `RefUnwindSafe` is not implemented for `(dyn MemoBuilder + Send + Sync + 'static)`
    = note: required because it appears within the type `*const (dyn MemoBuilder + Send + Sync + 'static)`
    = note: required because it appears within the type `Unique<(dyn MemoBuilder + Send + Sync + 'static)>`
    = note: required because it appears within the type `Box<(dyn MemoBuilder + Send + Sync + 'static)>`
    = note: required because it appears within the type `std::option::Option<Box<(dyn MemoBuilder + Send + Sync + 'static)>>`
    = note: required because it appears within the type `TransactionBuilder<FogResolver>`
    = note: required because it appears within the type `std::option::Option<TransactionBuilder<FogResolver>>`
    = note: required because of the requirements on the impl of `UnwindSafe` for `mc_util_ffi::FfiMutPtr<'_, std::option::Option<TransactionBuilder<FogResolver>>>`
    = note: required because it appears within the type `[closure@libmobilecoin/src/transaction.rs:343:40: 383:6]`

```

which I think cannot be resolved if we want to have `Box<dyn ...>` in the transaction builder, which I think we do.

Reply:
@the-real-adammork ^